### PR TITLE
Improve modularity and encapsulation

### PR DESCRIPTION
**Pull request by mpdude**
I have been trying to integrate PL (PHP/Twig-based) with an existing project. My goal is to use PL as a styleguide generator and be able to document existing Twig templates, macros and includes.

One issue I am facing is that currently PL does not work well when it is installed as a "pure" Composer dependency. That is when all parts like`core`, `patternengine-twig` and `styleguidekit-assets-default` end up next to each other in `vendor/pattern-lab/...`.

Regarding `styleguidekit-assets-default`, I'd like to suggest changing two things:

1) `gulpfile.js` should only build the assets from `src` to `dist`, but not make any assumptions about where the final styleguide lives.

More precisely, there are a lot of statements like `.pipe(gulp.dest('../../../public/styleguide/css'))`. Those break out of the `styleguidekit-assets-default` module. 

Keeping the `dist` folder and the final destination in sync should be part of the layer/code that integrates both.

2) Build `dist` in a way that it can simply be copied over to the final (PatternLab) styleguide.

If I am not mistaken, the final styleguide results from merging the files from this repo and the generated patterns/`.js` files in the `public` folder. Things would be much easier if I could just copy `dist` over to `public` and be set.

This means avoiding complicated re-arrangements that are currently expressed in `composer.json` `extra` data, which also requires the corresponding code (from PL `core`?) to be run at installation.

### CODE REVIEW

**Comment by dmolsen** [comment]
StyleguideKit is the "viewer" assets. So it's primarily the JS and CSS that makes up the iFrame that wraps the list of patterns and individual patterns (e.g. top bar nav, pop-up modals). Regarding your suggestions:

1) I use `gulpfile.js` as I develop and refine the viewer. I'm really not sure why it'd be used outside that context. Apart from customizing your own StyleguideKit and then... well, you could do whatever you like and wouldn't require a change to this specific repo. An "edition," which is the packaged version of Pattern Lab's dependencies, uses [Composer events](https://github.com/pattern-lab/edition-php-twig-standard/blob/662dff0c319bcbc3bf749d6c308a5018eb11b162/composer.json#L34-L50) to install items from `dist/` to their final location so I guess that is what keeps "the `dist` folder and the final destination in sync." Outside of my personal development environment node is _not_ a requirement at any point for Pattern Lab PHP.

2) I'll try make it a little cleaner to match the final output. That's a decent point. There's still the small issue of needing two directories. One for `source` and one for `public`. So it doesn't 100% address your needs.

I have a feeling you'll run into a lot more issues with your set-up. PL is going to expect the `core/` and `config/` directories that are found in the editions. Though maybe you've already addressed this. I can look at cleaning up the `packages/` folder requirement. I'm probably swimming upstream on that one and it's a personal preference.

**Comment by mpdude** [comment]
Dave, 

no matter if you use PatternLab `core`, `-php`, the Node variant or maybe some custom built solution (as I might need to build for Symfony based projects): They all would benefit if they could simply re-use `styleguidekit-assets-default`, as you guys have done a :+1: job here and we should not underestimate the amount of work and thought necessary to build this presentation front-end alone.

Reuse would work best if we'd just need to copy the `dist` folder (as-is) over to some docroot or maybe serve it to browsers directly. If I am not missing anything, all it needs in addition is a `patternlab-data.json` file that contains the menu structure, URLs for individual patterns and maybe other stuff. (I'll open another case so we can document that.)

That way, all projects (PL core/PHP/Node/...) could contribute and use improvements in this repo, although all of them may have different approaches to creating the `patternlab-data.json` and/or the pattern files.

Regarding the `gulpfile.js`, it would be great if everyone could use it as-is if they want to tweak the styleguidekit. It should be the tool to transform `src` into `dist` (if you happen to want to change `src` at all), but no more – that is, not try to copy things to the outside of this repo as it cannot know about the "context" the styleguidekit is used in.

**Comment by mpdude** [comment]
@dmolsen I have turned this into a PR that contains the changes I think are necessary.

This cannot be merged right now, as I started work off the latest v0.6.5 tag in order to get results I could compare with those of the other PL editions.

It would be great if you could check whether this state still works with your current process and the "editions": If things are copied based on the `composer.json` extra data, chances are everything is still OK even though the layout in `dist` changed.

**Comment by dmolsen** [comment]
@mpdude -

I appreciate the PR for reorganizing things. That said, I don't appreciate that you've determined that _my_ development workflow is wrong, has to change, and gone so far as to make changes that actively break it. You could have given me an out that supported my workflow and that was going to be my suggestion this morning. Give me a task that copies from dist to public so I can easily work on this in the development edition and I'll merge this stuff in. Otherwise I'll follow your lead with re-organizing this and then skip merging the PR.

**Comment by mpdude** [comment]
~~Hey @dmolsen,~~ 

~~sorry, I did not intend to break your workflow! I was hoping that the changes were transparent :-(~~

~~What exactly is wrong?~~

**Comment by mpdude** [comment]
Ah, I see.

It's that you want to run the `gulp` (or `gulp watch`) inside the `styleguidekit-assets-default` which is part of a larger *edition* you have installed. 

So, you need changes in `src` not only to update `dist` but also end in the `public` folder somewhere else. May I ask if you're running another build tool (or maybe LiveReload or the like) over in the *edition*?

**Comment by mpdude** [comment]
@dmolsen,

I added a `--copy-dist` argument to the `gulpfile.js` that you can use to copy all files not only to `dist` but also to another directory of your choice. 

`--copy-dist ../../../public` should make things work as before, with one exception (for the moment):

The `build:css-custom` task would previously copy _not_ to `../../../public` but to `../../../source/styleguide/css`.

Could you please explain what is so special about this one so I can see how we could address this?

Thanks!

**Comment by bmuenzenmeyer** [comment]
It's also worth noting to me that I am currently integrating PL node with this repo and any changes made here will affect that. I guess I'll stay tuned.

**Comment by mpdude** [comment]
@bmuenzenmeyer that's great news! Do you consider these changes as helpful as well?

**Comment by dmolsen** [comment]
@mpdude -

I think for now we'll leave `build:css-custom` out needing to move. I think it's going to end up getting changed as we finalize some clean-up work. I can add it back. The key stuff from my own development perspective is stuff going to `./public`. I appreciate the addition of the task. This will be merged to `dev` and `master` this weekend.

@bmuenzenmeyer -

I think @mpdude made a good point that I was making it way more complicated than it had to be. `./dist` should have been a 1-to-1 on what "production" will be for each of us (which is the `./public` layout). This reorganization will be done this weekend. `dev` still needs to bake. I have to add a mechanism for front-end plugins to modify tabs that can be found in the style guide UI in `dev`. That's basically my last big hurdle for the front-end.

**Comment by bmuenzenmeyer** [comment]
> @bmuenzenmeyer that's great news! Do you consider these changes as helpful as well?

Probably. I am doing some translation upon consumption, but it wasn't too bad overall

**Comment by mpdude** [comment]
@bmuenzenmeyer  What kind of translations?

File: gulpfile.js [hunk 1]
```diff
@@ -11,84 +11,92 @@ plugins.del            = require("del");
...
 
 gulp.task('build:css-custom', ['clean:css-custom'], function() {
 	return plugins.rubySass('src/sass/styleguide-specific.scss', { style: 'expanded', "sourcemap=none": true })
 		.pipe(plugins.autoprefixer({browsers: ['last 2 version', 'safari 5', 'ie 8', 'ie 9', 'opera 12.1', 'android 4']}))
-		.pipe(gulp.dest('dist/css/custom'))
+		.pipe(gulp.dest('dist/styleguide/css/custom'))
 		.pipe(plugins.rename({suffix: '.min'}))
 		.pipe(plugins.minifyCss())
-		.pipe(gulp.dest('dist/css/custom'))
-		.pipe(gulp.dest('../../../source/styleguide/css'));
```
Comment (mpdude) [inline gulpfile.js:79]: @dmolsen This also seems to be a "one-off" or special case that we might need to address?

Comment (mpdude) [inline gulpfile.js:79]: Misread Dave's last post, disregard.

**Comment by bmuenzenmeyer** [comment]
I've come to more fully realize the value of what @mpdude is proposing. 

A lot of my remaining integration problems seem to revolve around shuffling these assets correctly in place. For example, running `gulp` on the feature-panels branch repo of styleguidekit-assets-default wrote some files directly to the C:/ drive when attempting to update the dist assets on that branch. My hope was to update these independently after some recent `src/` changes.

Installing this repo via npm as part of [edition-node-gulp](https://github.com/pattern-lab/edition-node-gulp/blob/dev/package.json#L13)) is a huge goal of the node port, and it seems as if the command line arg this PR proposes would do the trick, making npm consumption of 'dist/' very simple.

I cannot speak for the entirety of the PR, especially the composer.json file, but am 100% in favor of these gulpfile changes.

cc @dmolsen

**Comment by dmolsen** [comment]
This has been merged into `feature-panels`. It will eventually make it's way to `master` once the updates to how panels are handled are complete.

**Pull request closed without merging**